AT fails to build on RHEL 8

I've now tried to build on two RHEL 8.1 systems, and they both fail identically:
```
...
Starting the build of gcc_2...
make: *** [Makefile:1342: /home/pc/advance-toolchain/at14.0-0-alpha.redhat-8_ppc64le_ppc64le/receipts/gcc_2.b.rcpt] Error 1
```
`at14.0-0-alpha.redhat-8_ppc64le_ppc64le/logs/_gcc_2-3_standard_buildf-06_make.log`:
```
configure: error: in `/home/pc/advance-toolchain/at14.0-0-alpha.redhat-8_ppc64le_ppc64le/builds/gcc_2/libiberty':
configure: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
checking for string.h... /opt/at-next-14.0-0-alpha/bin/gcc -E
yes
checking for sys/stat.h... make[3]: *** [Makefile:10923: configure-stage1-libiberty] Error 1
```
`at14.0-0-alpha.redhat-8_ppc64le_ppc64le/builds/gcc_2/libiberty/config.log`:
```
configure:3495: checking whether we are cross compiling
configure:3503: /opt/at-next-14.0-0-alpha/bin/gcc -o conftest -g  -Wl,--dynamic-linker=/opt/at-next-14.0-0-alpha/lib64/ld64.so.2  conftest.c  >&5
configure:3507: $? = 0
configure:3514: ./conftest
/home/pc/advance-toolchain/at14.0-0-alpha.redhat-8_ppc64le_ppc64le/sources/gcc/libiberty/configure: line 3516: 103800 Segmentation fault      ./conftest$ac_cv_exeext
configure:3518: $? = 139
configure:3525: error: in `/home/pc/advance-toolchain/at14.0-0-alpha.redhat-8_ppc64le_ppc64le/builds/gcc_2/libiberty':
configure:3527: error: cannot run C compiled programs.
```
And indeed, the newly built linker does not produce working executables:
```
$ echo 'int main(){}' > conftest.c
$ /opt/at-next-14.0-0-alpha/bin/gcc -o conftest -g  -Wl,--dynamic-linker=/opt/at-next-14.0-0-alpha/lib64/ld64.so.2  conftest.c
$ ./conftest
Segmentation fault (core dumped)
```
It seems to be caused by using the newly built loader, the newly built exectable, and the _system_ libc:
```
$ ldd ./conftest
        linux-vdso64.so.1 (0x00007fff85e00000)
        libc.so.6 => /lib64/power9/libc.so.6 (0x00007fff85bd0000)
        /opt/at-next-14.0-0-alpha/lib64/ld64.so.2 => /lib64/ld64.so.2 (0x00007fff85e20000)
```
Manually adding an `rpath` which finds the newly built libc instead produces a working executable:
```
$ /opt/at-next-14.0-0-alpha/bin/gcc -o conftest -g  -Wl,--dynamic-linker=/opt/at-next-14.0-0-alpha/lib64/ld64.so.2  conftest.c -Wl,-rpath=/opt/at-next-14.0-0-alpha/lib64
$ ldd ./conftest
        linux-vdso64.so.1 (0x00007fffbdf90000)
        libc.so.6 => /opt/at-next-14.0-0-alpha/lib64/libc.so.6 (0x00007fffbdd70000)
        /opt/at-next-14.0-0-alpha/lib64/ld64.so.2 => /lib64/ld64.so.2 (0x00007fffbdfb0000)
$ ./conftest
$ 
```

I've found it challenging to fully understand the process.
As Tulio said, `--with-advance-toolchain` is used for GCC stage 2 build. However, it is only applied in `gcc/config.gcc` (by `atcfg_pre_hacks` and `atcfg_configure` in `configs/13.0/packages/gcc/stage_2`). This may be too late, as other modules are built before `gcc` module. Here, `intl` fails:
```
make[3]: Entering directory '/home/pc/at13/at14.0-0-alpha.redhat-8_ppc64le_ppc64le/builds/gcc_2'
Configuring stage 1 in ./intl
configure: creating cache ./config.cache
checking for powerpc64le-linux-gnu-gcc... /opt/at-next-14.0-0-alpha/bin/gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... configure: error: in `/home/pc/at13/at14.0-0-alpha.redhat-8_ppc64le_ppc64le/builds/gcc_2/intl':
configure: error: cannot run C compiled programs.
```

--- a/scripts/helpers/utilities.mk
+++ b/scripts/helpers/utilities.mk
@@ -364,6 +364,19 @@ endef
 # still distributed in /lib or /usr/lib instead of the correct multiarch
 # directory.  So, it's necessary to carry a hack until they complete to port
 # all packages to multiarch.
+#
+# Another hack below is the "syslib-override" directory.
+# During the build, there is an awkward period where some packages are
+# built depending on AT libraries, but optimized libraries are not yet
+# built. The loader will always prefer optimized libraries *even if they
+# are under directories lower in the search order*.  For example,
+# /lib64/power9/libc.so is preferred over /opt/atX.0/lib64/libc.so.
+# This causes all sorts of problems during this awkward period.
+# So, the syslib-override directory is put at the top of the search order
+# and populated with AT-built (not-optimized) libraries in optimized
+# subdirectories, so they will be used in preference, and the build
+# can proceed through the awkward period. After that period, the entire
+# syslib-override directory tree is removed.
 define prepare_loader_cache
     set -e; \
     if [[ ! -d "$(AT_DEST)/etc/ld.so.conf.d" ]]; then \
@@ -377,7 +390,11 @@ define prepare_loader_cache
         echo "/usr/local/lib"                 >> "$(DYNAMIC_LOAD)/sys32.ld.conf"; \
     fi; \
     if [[ -n "$(TARGET64)" ]]; then \
-        echo "$(AT_DEST)/$(TARGET)/lib64" >  "$(DYNAMIC_LOAD)/at64.ld.conf"; \
+        > "$(DYNAMIC_LOAD)/at64.ld.conf"; \
+        if [[ $1 -eq 1 ]]; then \
+            echo "$(AT_DEST)/syslib-override"   >  "$(DYNAMIC_LOAD)/at64.ld.conf"; \
+        fi; \
+        echo "$(AT_DEST)/$(TARGET)/lib64" >> "$(DYNAMIC_LOAD)/at64.ld.conf"; \
         echo "$(AT_DEST)/lib64"           >> "$(DYNAMIC_LOAD)/at64.ld.conf"; \
         echo "/lib64"                     >  "$(DYNAMIC_LOAD)/sys64.ld.conf"; \
         if [[ ("$(DISTRO_FM)" == "ubuntu") || ("$(DISTRO_FM)" == "debian") ]]; then \
@@ -421,6 +438,17 @@ define prepare_loader_cache
         cat "$(DYNAMIC_LOAD)/sys32.ld.conf" >> "$(DYNAMIC_LOAD)/ld.so.conf"; \
     fi; \
     grep -E -v "^$$" "$(DYNAMIC_LOAD)/ld.so.conf" > "$(AT_DEST)/etc/ld.so.conf"; \
+    if [[ $1 -eq 2 ]]; then \
+        rm -rf $(AT_DEST)/syslib-override; \
+    else \
+        for cpu in $(BUILD_ACTIVE_MULTILIBS); do \
+            mkdir -p $(AT_DEST)/syslib-override/"$${cpu}"; \
+            (\cd $(AT_DEST)/lib64 && \
+             find . -maxdepth 1 -type f \
+                  -exec cp -p '{}' $(AT_DEST)/syslib-override/"$${cpu}"/'{}' \; \
+            ); \
+        done; \
+    fi; \
     "$(AT_DEST)/sbin/ldconfig"; \
     group=$$( ls $(DYNAMIC_SPEC)/ | grep "toolchain\$$" ); \
     echo "$(AT_DEST)/etc/ld.so.conf" \

